Blog: MQTT: The common pitfalls in industrial data operations (#2709)
FlowFuse/website: src/blog/2024/11/mqtt-common-pitfalls-industrial-data-operations.md

---
title: "Common Pitfalls of Using MQTT in Industrial Data Operations"
subtitle: We explore 6 Common Pitfalls of using MQTT in Industrial Data Operations and how you can overcome them
description: Learn about the common pitfalls of using MQTT in industrial data operations, including data quality issues, security risks, and vendor lock-in.

> date: 2024-11-22

Reviewer: ```suggestion
date: 2024-11-25
```

authors: ["sumit-shinde"]
image: /blog/2024/11/images/mqtt-common-pitfalls.png
keywords: mqtt disadvantages, mqtt pitfalls, mqtt challenges, mqtt reliability problems, mqtt drawbacks, mqtt limitations, cons of using mqtt
tags:
   - post
   - flowfuse
   - mqtt
---

In industrial environments, where real-time data is essential for smooth operations, MQTT has become a go-to messaging protocol for connecting devices. But like any technology, it’s not without its challenges. From data quality issues to security risks and vendor lock-in, there are several pitfalls that can disrupt your industrial data flow. In this article, we will discuss six common MQTT challenges and explore strategies and solutions to help you mitigate these issues, ensuring more reliable, secure, and efficient industrial data operations.

<!--more-->

> The challenges we’ll discuss depend largely on an organization’s specific use cases. While each of these obstacles can have benefits in certain scenarios, they also present challenges that could hinder smooth data operations. By identifying these pitfalls early, you can take proactive steps to improve data communication and better manage your industrial processes.

Reviewer: Good intro @sumitshinde-84 , thanks.

### Lack of Standardization in Data Formats and Quality Concerns

> In industrial environments, leveraging data effectively is critical for informed decision-making and operational efficiency. This data often comes from a variety of IoT devices, but for it to be truly useful, it must be standardized and formatted in a way that makes it easily processable by the receiving systems.

Reviewer: ```suggestion
Industrial data often comes from a variety of devices, machines, and software services. For it to be truly useful, it must be standardized and formatted in a way that makes it easily processable by the receiving systems.
```

> A significant issue with MQTT is its **lack of enforced standardization** in data formats. While this flexibility can be advantageous in some contexts, it often means that additional work is required to transform data into a usable format, which can create inefficiencies. Without a standardized approach, data may become inconsistent or poorly structured, which complicates analysis and makes it harder to extract actionable insights.

Reviewer: ```suggestion
A significant challenge with MQTT is its **lack of enforced standardization** in data formats. While this flexibility can be advantageous in some contexts, it often means that additional work is required to transform data into a usable format, which can create inefficiencies. Without a standardized approach, data may become inconsistent or poorly structured, which complicates analysis and makes it harder to extract actionable insights.
```

Reviewer: Furthermore, a line about how the consumer of the data doesn't know what to expect withouth message schema enforcement is another argument to add.

To address this, frameworks like **Sparkplug B** provide a standardized way of managing MQTT payloads, ensuring data consistency and quality across devices and systems. Leveraging such frameworks ensures that data is delivered in a structured, reliable format, which ultimately supports more informed decision-making and streamlined operations.

### Payload Limitations and Bandwidth Constraints

MQTT can struggle when transmitting large amounts of data or high-frequency messages, especially when bandwidth is limited. Although the MQTT standard allows message payloads up to 250 MB, sending such large amounts of data can cause delays and performance issues. This is because MQTT was designed for small, lightweight messages, and many brokers set limits on payload sizes. Large payloads, like images or detailed sensor data, can make this even worse.

> To avoid these problems, consider compressing your data before sending it, sending data less often by aggregating it, or adjusting the frequency of updates. By managing the size and frequency of messages and designing your topics efficiently, you can keep data flowing smoothly, even in bandwidth-limited environments.

Reviewer: When large files have to be send, please advise to store it on a shared storage layer like AWS S3 or network drives and sending a pointer to file. See also: https://flowfuse.com/blog/2024/01/unified-namespace-when-not-to-use/#large-files-or-binary-data

### Risks of Data Loss and Delivery Issues

> In industrial settings, the timely and accurate transmission of data is critical. Data loss, duplication, or out-of-order delivery can significantly impact decision-making, safety, and efficiency. Unfortunately, MQTT’s architecture — based on a publish-subscribe model — can lead to issues like message loss or misordering.

Reviewer: What do you mean with accurate transmission? Accuracy has to do with the subscriber being the actual subscriber. While I've heard about message loss on some brokers -- accuracy isn't in question. Can we change the adjective to for example reliability focussed?

Network disruptions, for example, can cause important messages to be dropped before they reach their destination. Similarly, when a network reconnects after a failure, MQTT might deliver messages out of order or resend some messages, creating duplicates. This can lead to confusion, unreliable analyses, or even unsafe conditions (e.g., in the case of robotic control systems, where mis-sequenced commands can lead to malfunction).

To mitigate these risks, MQTT offers **Quality of Service (QoS) levels**, which define how messages are delivered:

- **QoS 0**: Messages are delivered at most once with no guarantees of delivery or order.
- **QoS 1**: Guarantees that messages are delivered at least once, but duplicates can occur.
- **QoS 2**: Ensures messages are delivered exactly once, and in the correct order.

While higher QoS levels offer more reliable message delivery, they also introduce increased overhead and latency, which may not be acceptable in real-time applications. Therefore, it’s critical to evaluate which QoS level best aligns with the operational needs and time sensitivity of your industrial processes.

### Lack of Acknowledgment Mechanisms

> Another key challenge with MQTT is the **lack of acknowledgment mechanisms**. Unlike traditional request-response communication models, MQTT doesn’t inherently provide a built-in way for the receiver to acknowledge receipt of a message. This lack of visibility can make it difficult to ensure that data has been successfully transmitted, which could lead to uncertainties in operational systems.

Reviewer: ```suggestion
Another key challenge with MQTT and other Pub/Sub architectures, is the **lack of acknowledgment mechanisms**. Unlike traditional request-response communication models, MQTT doesn’t inherently provide a built-in way for the receiver to acknowledge receipt of a message. This lack of visibility can make it difficult to ensure that data has been successfully transmitted, which could lead to uncertainties in operational systems.
```

While MQTT allows for some workarounds (such as implementing tracking mechanisms or using a second topic for acknowledgment), these solutions add complexity to the integration process. By designing systems that include these acknowledgment mechanisms, you can improve reliability, but it requires careful planning and additional development effort.

### TCP Limitations in MQTT Frameworks

> While MQTT operates over the TCP/IP protocol to ensure reliable message delivery, this introduces several challenges, particularly for resource-constrained industrial devices.

Reviewer: TCP/IP doesn't operate on a message level, but on a packet level: https://en.wikipedia.org/wiki/OSI_model.

```suggestion
MQTT operates over the TCP/IP protocol to ensure connections between clients and the broker, which introduces several challenges for resource-constrained industrial devices.
```

> TCP requires maintaining a constant connection state, which demands significant processing power and memory. This becomes problematic in industrial IoT environments, where many devices are low-powered and designed for minimal energy consumption. The ongoing need for devices to manage and maintain a persistent TCP connection can quickly drain resources, limiting overall system efficiency and scalability.

Reviewer: Industrial IoT I've not seen struggle with lack of powerful devices. It depends on what level of the 

```suggestion
TCP requires maintaining a constant connection state, which demands significant processing power and memory. This becomes problematic for some IoT environments, especially on the field level, where many devices are low-powered and designed for minimal energy consumption. The ongoing need for devices to manage and maintain a persistent TCP connection can quickly drain resources, limiting overall system efficiency and scalability.
```

Moreover, not all IoT devices are capable of supporting the overhead required by TCP/IP. Simpler devices, such as sensors with limited computing capabilities, may struggle with the complexities of maintaining a reliable TCP connection. This can restrict the deployment of MQTT in systems where lightweight communication protocols, like UDP, would be more appropriate for ensuring broad compatibility and energy efficiency.

### Security Risks and Vulnerabilities

> Security is a critical concern when using MQTT in industrial environments. Since MQTT does not include built-in encryption or authentication, it can be vulnerable to various attacks such as data interception, unauthorized access, or man-in-the-middle attacks. In industrial settings, this could result in the manipulation of data, disruption of operations, or even cause safety issues with critical systems.

Reviewer: I think you got this wrong -- it's possible to missconfigure MQTT to not require authentication. Like you say below, it's possible to roll it out securely.
